src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperDirectiveSpanVisitor.cs

```diff
@@ -82,50 +83,42 @@ public IEnumerable<TagHelperDescriptor> GetDescriptors(Block root)
 
         public override void VisitSpan(Span span)
         {
-            // We're only interested in spans with an AddOrRemoveTagHelperChunkGenerator.
+            if (span == null)
+            {
+                throw new ArgumentNullException(nameof(span));
+            }
 
-            if (span.ChunkGenerator is AddOrRemoveTagHelperChunkGenerator)
+            TagHelperDirectiveType directiveType;
+            if (span.ChunkGenerator is AddTagHelperChunkGenerator)
+            {
+                directiveType = TagHelperDirectiveType.AddTagHelper;
+            }
+            else if (span.ChunkGenerator is RemoveTagHelperChunkGenerator)
             {
-                var chunkGenerator = (AddOrRemoveTagHelperChunkGenerator)span.ChunkGenerator;
-
-                var directive =
-                    chunkGenerator.RemoveTagHelperDescriptors ?
-                    TagHelperDirectiveType.RemoveTagHelper :
-                    TagHelperDirectiveType.AddTagHelper;
-                var textLocation = GetSubTextSourceLocation(span, chunkGenerator.LookupText);
-
-                var directiveDescriptor = new TagHelperDirectiveDescriptor
-                {
-                    DirectiveText = chunkGenerator.LookupText,
-                    Location = textLocation,
-                    DirectiveType = directive
-                };
-
-                _directiveDescriptors.Add(directiveDescriptor);
+                directiveType = TagHelperDirectiveType.RemoveTagHelper;
             }
             else if (span.ChunkGenerator is TagHelperPrefixDirectiveChunkGenerator)
             {
-                var chunkGenerator = (TagHelperPrefixDirectiveChunkGenerator)span.ChunkGenerator;
-                var textLocation = GetSubTextSourceLocation(span, chunkGenerator.Prefix);
-
-                var directiveDescriptor = new TagHelperDirectiveDescriptor
-                {
-                    DirectiveText = chunkGenerator.Prefix,
-                    Location = textLocation,
-                    DirectiveType = TagHelperDirectiveType.TagHelperPrefix
-                };
-
-                _directiveDescriptors.Add(directiveDescriptor);
+                directiveType = TagHelperDirectiveType.TagHelperPrefix;
+            }
+            else
+            {
+                // Not a chunk generator that we're interested in.
+                return;
             }
-        }
 
-        private static SourceLocation GetSubTextSourceLocation(Span span, string text)
-        {
-            var startOffset = span.Content.IndexOf(text);
+            var directiveText = ((TagHelperDirectiveChunkGenerator)span.ChunkGenerator).GetDirectiveText(span);
```

Comment:
Also the `(TagHelperDirectiveChunkGenerator)` cast is fragile. This class assumes `AddTagHelperChunkGenerator` and so on share this common base. That's not information that should need to be shared.